tests/errors/cdef_multiple_inheritance_subclassing_check.pyx

```diff
@@ -0,0 +1,80 @@
+# mode: run
+
+# https://github.com/cython/cython/issues/4350
+# In some circumstances it's possible to multiply inherit Python classes
+# from cdef classes in ways that cause __cinit__ and __dealloc__ not to
+# be called.
+
+# The tests work on Python<3.6 because Cython itself calls __init_subclass__.
+# If they were written using exec then they would be fail there.
+
+cdef class BaseA:
+    cdef int x
+
+    def __cinit__(self):
+        print "BaseA.__cinit__"
+
+cdef class A(BaseA):
+    # layout compatible with (and derived from) A
+    def __cinit__(self):
+        print "A.__cinit__"
+
+cdef class X:
+    def __cinit__(self):
+        print "X.__cinit__"
+
+cdef class Y:
+    # layout compatible with X
+    def __cinit__(self):
+        print "Y.__cinit__"
+
+class BaseC(BaseA):
+        pass
+
+def test_BaseC_A():
+    """
+    >>> test_BaseC_A()  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+        ...
+    TypeError: Invalid inheritance from cdef class A: ...
+    """
+    class C(BaseC, A):
+        pass
+
+    C()
+
+def test_A_BaseC():
+    """
+    Should work so we just test the constructors are right
+    >>> test_A_BaseC()
+    BaseA.__cinit__
+    A.__cinit__
+    """
+    class C(A, BaseC):
+        pass
+
+    C()
+
+def test_X_Y():
+    """
+    >>> test_X_Y() # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+        ...
+    TypeError: Invalid inheritance from cdef class Y: ...
+    """
+    class C(X, Y):
```

Comment:
Potentially we could slacken the rules on this case (inheriting from two empty cdef classes) iff the offending class has a trivial `tp_new` and `tp_dealloc` (i.e. has no user defined `__cinit__` or `__dealloc__` and Cython doesn't need to generate anything special). Even a single `cdef` function would cause it not to apply (because of the vtab)

I'm not convinced that it's worth the effort to detect, but we could do it in future if it generates a bunch of "you broke my code" issues.